piper/ledbutton.py

```diff
@@ -0,0 +1,181 @@
+        """The selected color, as Gdk.RGBA."""
+        return self.colorbutton.get_rgba()
+
+    @GObject.Property
+    def brightness(self):
+        """The selected brightness value, as int."""
+        return self.adjustment_brightness.get_value()
+
+    @GObject.Property
+    def effect_rate(self):
+        """The selected effect rate, as int."""
+        return self.adjustment_effect_rate.get_value()
+
+
+@GtkTemplate(ui="/org/freedesktop/Piper/ui/LedButton.ui")
+class LedButton(Gtk.ButtonBox):
+    """A Gtk.ButtonBox subclass to implement the toggle buttons that set an
+    LED's mode."""
+
+    __gtype_name__ = "LedButton"
+
+    button_solid = GtkTemplate.Child()
+    button_cycle = GtkTemplate.Child()
+    button_breathing = GtkTemplate.Child()
+    button_off = GtkTemplate.Child()
+
+    def __init__(self, ratbagd_led, *args, **kwargs):
+        """Instantiates a new LedButton.
+
+        @param ratbagd_led The LED to configure, as ratbagd.RatbagdLed
+        """
+        Gtk.ButtonBox.__init__(self, *args, **kwargs)
+        self.init_template()
+        self._led = ratbagd_led
+        self._init_ui()
+
+    def _init_ui(self):
+        # Activate the button that maps to the LED's current mode
+        mode = self._led.mode
+        if mode == RatbagdLed.LED_MODE_OFF:
+            self.button_off.set_active(True)
+        elif mode == RatbagdLed.LED_MODE_ON:
+            self.button_solid.set_active(True)
+        elif mode == RatbagdLed.LED_MODE_CYCLE:
+            self.button_cycle.set_active(True)
+        elif mode == RatbagdLed.LED_MODE_BREATHING:
+            self.button_breathing.set_active(True)
+
+        # Only connect to the signals now so we don't map a dialog when the main
+        # window isn't up yet.
+        self.button_off.connect("clicked", self._on_button_clicked, RatbagdLed.LED_MODE_OFF)
+        self.button_solid.connect("clicked", self._on_button_clicked, RatbagdLed.LED_MODE_ON)
+        self.button_cycle.connect("clicked", self._on_button_clicked, RatbagdLed.LED_MODE_CYCLE)
+        self.button_breathing.connect("clicked", self._on_button_clicked, RatbagdLed.LED_MODE_BREATHING)
+
+    def _on_button_clicked(self, button, mode):
+        # Responds to a click on the button. The button internally always
+        # toggles on a click, so we cannot toggle only when the user applies
+        # changes in the dialog and abort when the user cancels the dialog. As
+        # such, we always change the mode; this is probably fine since the user
+        # has to commit the changes to the device anyway, but it's not ideal.
+        if not button.get_active():
+            # Prevent the unactivated togglebutton's dialog from showing.
+            return
+
+        self._led.mode = mode
+        if mode == RatbagdLed.LED_MODE_ON:
+            self._open_solid_dialog()
+        elif mode == RatbagdLed.LED_MODE_CYCLE or mode == RatbagdLed.LED_MODE_BREATHING:
+            self._open_advanced_dialog(mode)
```

Comment:
cache the currently active one and get the result of the dialog check for cancel. If cancel, you could re-toggle the currently active one? Not sure whether that'll flicker though.